### PR TITLE
Generate: Relaxed `max_length` and `max_new_tokens` coexistence

### LINKED ISSUE #21369: "Both `max_new_tokens` and `max_length` have been set but they serve the same purpose" when only setting max_new_tokens.
### System Info

- `transformers` version: 4.27.0.dev0
- Platform: Windows-10-10.0.19044-SP0
- Python version: 3.10.4
- Huggingface_hub version: 0.12.0
- PyTorch version (GPU?): 1.13.1+cpu (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@gante I believe.

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

1. Install transformers
2. Run the following script 
```
from transformers import pipeline

summarizer = pipeline("summarization", max_new_tokens=50)

prompt = "text to summarize"
summary = summarizer(prompt)
```
3. It crashes with the following error:

> ValueError: Both `max_new_tokens` and `max_length` have been set but they serve the same purpose -- setting a limit to the generated output length. Remove one of those arguments. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)

### Expected behavior

Expected behavior is that the script should run. I only set `max_new_tokens` and not `max_length`.
This might seem like a duplicate of https://github.com/huggingface/transformers/issues/20894 but the issue persists even after installing the latest version of transformers with the fix for that issue.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

TL;DR: stops raising an exception in `.generate()` when `max_length` and `max_new_tokens` are both set -- `max_new_tokens` will take precedence.

Context: Some downstream uses of `.generate()`, for legacy reasons, set `max_length` (e.g. pipelines, API). If a user tries manually setting `max_new_tokens`, as suggested in the documentation, an exception is thrown (even if `max_length` is manually set to `None`). 

Because `max_length` can be set outside the `GenerationConfig` and `.generate()`, it's hard to detect whether the `max_length` is intentionally set (and thus shouldn't be allowed together with `max_new_tokens`) or simply a helpful default. Raising an exception can thus block a correct usage of `max_new_tokens`. This PR relaxes this requirement, making `max_new_tokens` take precedence and raising an informative warning instead.

Fixes #21369 

___________________________________________________________________
Example of failing code before this PR:
```py

import requests

API_URL = "https://api-inference.huggingface.co/models/google/flan-t5-xl"
headers = {"Authorization": "Bearer hf_xxx"}


def query(payload):
    response = requests.post(API_URL, headers=headers, json=payload)
    return response.json()


output = query(
    {"inputs": "The answer to the universe is", "parameters": {"max_new_tokens": 100}}
)
output
```

### CODE REVIEW

**Comment by gante** [comment]
cc @lewtun -- thank you for raising the issue 🙏

File: src/transformers/generation/tf_utils.py [hunk 1]
```diff
@@ -700,21 +700,21 @@ def generate(
...
             generation_config.max_length = generation_config.max_new_tokens + input_ids_seq_length
-        elif not has_default_max_length and generation_config.max_new_tokens is not None:
-            raise ValueError(
-                "Both `max_new_tokens` and `max_length` have been set but they serve the same purpose -- setting a"
-                " limit to the generated output length. Remove one of those arguments. Please refer to the"
-                " documentation for more information. "
-                "(https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)"
-            )
+            if not has_default_max_length:
+                warnings.warn(
```
Comment (patrickvonplaten) [inline src/transformers/generation/tf_utils.py:711]:
```suggestion
                logger.warn(
```

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1274,21 +1274,21 @@ def generate(
...
             generation_config.max_length = generation_config.max_new_tokens + input_ids_seq_length
-        elif not has_default_max_length and generation_config.max_new_tokens is not None:
-            raise ValueError(
-                "Both `max_new_tokens` and `max_length` have been set but they serve the same purpose -- setting a"
-                " limit to the generated output length. Remove one of those arguments. Please refer to the"
-                " documentation for more information. "
-                "(https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)"
-            )
+            if not has_default_max_length:
+                warnings.warn(
```
Comment (patrickvonplaten) [inline src/transformers/generation/utils.py:1285]:
```suggestion
                logger.warn(
```

**Review by patrickvonplaten — APPROVED** [review]
Thanks!

**Review by sgugger — APPROVED** [review]
LGTM, thanks for working on this!
